scene/main/viewport.cpp

```diff
@@ -476,6 +476,19 @@ void Viewport::_notification(int p_what) {
 #endif // _3D_DISABLED
 		} break;
 
+		case NOTIFICATION_INTERNAL_PROCESS: {
+#ifndef _3D_DISABLED
+			if (use_xr && XRServer::get_singleton() != nullptr) {
+				Ref<XRInterface> xr_interface = XRServer::get_singleton()->get_primary_interface();
+				if (xr_interface.is_valid() && xr_interface->is_initialized()) {
+					Size2 xr_size = xr_interface->get_render_target_size();
+					// This will only actually resize if the the size has changed.
+					_set_size((Size2i)xr_size, Size2i(), Rect2i(), Transform2D(), true);
```

Comment:
I think we should add viewcount to _set_size instead of force_resize, with force_resize we'll end up recreating all buffers every frame which would significantly slow things down. Getting the view count from the xr_interface here and pushing it to the rendering server will make things more stable.

Also if use_xr is true but there is no usable XR interface, we should set the viewport size to 0.0, this will free all buffers and skip rendering the viewport which is what we want if the XR interface has not been initialised yet.

Reply:
Ergo in `_set_size` the call to `RS::get_singleton()->viewport_set_size(viewport, size.width, size.height);` should become `RS::get_singleton()->viewport_set_size(viewport, size.width, size.height, view_count);`

Reply:
> I think we should add viewcount to _set_size instead of force_resize, with force_resize we'll end up recreating all buffers every frame which would significantly slow things down.

FYI, the `true` at the end isn't for `p_force_resize`, it's for `p_allocated`. I'm only passing `true` for `p_force_resize` in the `set_use_xr()` method.

However, I would also feel better tracking the number of views! It made me a little nervous not having that in there, so I'll happily add it :-)

> Also if use_xr is true but there is no usable XR interface, we should set the viewport size to 0.0, this will free all buffers and skip rendering the viewport which is what we want if the XR interface has not been initialised yet.

Sounds good!